### PR TITLE
Building a nodejs app with Docker

### PR DESCRIPTION
**Pull request by damikanye**
The tutorial is about running a simple Node.js application using Docker especially using a Dockerfile to build a Docker image.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
Hi, @damikanye. Thanks for submitting an article. It looks really interesting. I've used Node.js and Docker before but never in combination. 

Just a heads up, I've started a preliminary review. Do you have a hero image you'd like to use for the article (600x400px preferably) or would you prefer one to be sourced for you?

**Comment by louisefindlay23** [comment]
@hectorkambow, can you add some front-matter, please?

**Comment by damikanye** [comment]
I prefer it being sourced @louisefindlay23

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+})
+```
+
+### Run the application
+
+The app is ready to launch:
+```
+$ node app.js
+```
+Go to `http://localhost:8080/` in your browser to view it
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:82]: You should add full stops after sentences.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+First, we will start by creating a directory for our project then install some dependencies for our simple Hello World Website.
+```
+ mkdir node-web-app
+ cd node-web-app
+```
+
+### Install NPM and Express Framework
+
+Install npm and install the express framework, which is one of Node.js frameworks. Initialize npm in our directory.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:32]: Remove second install for readability. Which is **a** Node.js framework would also flow better as well as **Then,** initialize

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+```
+ npm init
+```
+NPM creates a `package.json` that holds the dependencies of the app. Install the express framework dependency.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:36]: The case (capitalisation) of NPM should be consistent. Same comment as before about Install Install. Consider using words such as then, next and finally to improve the flow.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  "license": "ISC",
+  "dependencies": {
+    "express": "^4.17.1"
+  },
+  "devDependencies": {},
+  "description": ""
+}
+
+```
+Create `app.js` file with an http server that will set up the Hello World site:
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:59]: http should be all caps.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+# Step 2 - Create A DockerFile
+
+Create a file in the directory called `Dockerfile`
+
+The first thing is to define which image we want to build from. Here we will use version `9` of `node` available from Docker Hub:
+```
+FROM node:9
+```
+Next is to create the working directory for your application
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:92]: Next should have a comma after it. Would flow better with Next, create a working directory

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+# Copying rest of the application to app directory
+COPY . /app
+ 
+# Expose the port and start the application
+Expose 8080
+
+CMD ["npm","start"]
+```
+*NB: Above you will notice we have two distinct COPY command was used to reduce the application rebuild time*
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:134]: Doesn't make sense. We used two distinct Copy commands to reduce

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+### .dockerignore file
+
+Create a `.dockerignore` file in the directory so as not to copy unneccessary files to the container:
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:138]: Remove in the directory to simplify the sentence.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+```
+node_modules
+npm-debug.log
+```
+This prevents local module and debugs logs from being copied onto your Docker image.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:143]: **the local module** and omit the **s** from debugs.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+# Step 3 - Building your Docker Image
+
+Building your Docker image is quiet easy and can be done using a simple command.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:147]: Should be quite not quiet. Single command would flow better.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+Use the `docker ps` command to check the running container.
+```
+docker ps -a
+```
+Result:
+```
+CONTAINER ID        IMAGE                        COMMAND                CREATED              STATUS                          PORTS                    NAMES
+34fe3040ff6b        cisca                        "npm start"            About a minute ago   Exited (1) About a minute ago        0.0.0.0:8080->8080/tcp                    intelligent_hofstadter
+```
+
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:234]: Perhaps create a conclusion heading?

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+This tutorial has helped you with the basics of running a simple Node.js application using Docker especially using a Dockerfile to build a Docker image.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:235]: Change especially to and would aid the flow.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Great article, @damikanye. I enjoying reading it. Just one or two minor spelling mistakes and grammar issues you should sort out. I also pointed out areas where you could improve readability and markdown formatting. 

Don't be discouraged, I'm looking forward to building my first Node.js app using a Dockerfile when I pass it onto tech review. Make the changes I've suggested (and scan the article for more cases of the issues I've highlighted) and it should be approved to the next review stage.

File: articles/building-a-nodejs-application-using-docker/index.md (hunk 10 above)
Comment (louisefindlay23) [reply articles/building-a-nodejs-application-using-docker/index.md:234]: You could add a heading here called Conclusion. `## Conclusion`

**Comment by louisefindlay23** [comment]
> 
> 
> I prefer it being sourced @louisefindlay23

I've added a hero image, @damikanye as requested. Let me know if you'd like me to change it.

**Comment by damikanye** [comment]
> > I prefer it being sourced @louisefindlay23
> 
> I've added a hero image, @damikanye as requested. Let me know if you'd like me to change it.

It's cool.

**Comment by hectorkambow** [comment]
Plagiarism check done

**Comment by louisefindlay23** [comment]
@damikanye, I've noticed you've made some changes. Are you ready for me to do another review?

**Comment by damikanye** [comment]
> @damikanye, I've noticed you've made some changes. Are you ready for me to do another review?

Yes

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+- Creating A Node.js Application
+- Create A DockerFile
+- Building your Docker Image
+- Expose
+
+# Step 1 - Creating a Node.js Application
+
+First, we will start by creating a directory for our project then install some dependencies for our simple Hello World Website.
+```Bash
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:25]: Bash needs to be lowercase

Comment (damikanye) [reply articles/building-a-nodejs-application-using-docker/index.md:25]: ohh..kay

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+- [MacOS](https://docs.docker.com/docker-for-mac/install/)
+
+In order to dockerize a Node.js Application we need to follow these steps:
+
+- Creating A Node.js Application
+- Create A DockerFile
+- Building your Docker Image
+- Expose
+
+# Step 1 - Creating a Node.js Application
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:22]: This needs to be a H2 heading

File: articles/building-a-nodejs-application-using-docker/index.md (hunk 12 above)
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md]: Bash needs to be lowercase

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
+# Building A Node.js Application Using Docker
+
+[Docker](https://www.docker.com/) allows us to run our applications as *containers*. A container is a standalone executable package which is lightweight and has everything needed to run an application be it libraries, tools, runtime, setting or code.
+
+In this tutorial, we will build a Node.js application, create an image and also build a container using the image. Enjoy!
+
+
+# Dockerize the Node.js Application
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:8]: This needs to be a H2 heading

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+### Run The Application
+
+The app is ready to launch:
+```
+$ node app.js
+```
+Go to `http://localhost:8080/` in your browser to view it.
+
+# Step 2 - Create A DockerFile
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:84]: This needs to be a H2 heading

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+COPY /app
+```
+Expose the port and start the application
+```
+Expose 8080
+
+CMD ["npm", "start"]
+```
+Your `Dockerfile` should look like this:
+```Bash
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md]: Lowercase

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+## .dockerignore file
+
+Create a `.dockerignore` file so as not to copy unneccessary files to the container:
+```
+node_modules
+npm-debug.log
+```
+This prevents the local module and debug logs from being copied onto your Docker image.
+
+# Step 3 - Building your Docker Image
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:145]: This needs to be a H2 heading

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+docker image ls
+```
+Result:
+```
+REPOSITORY                    TAG                 IMAGE ID            CREATED             SIZE
+cisca                         latest              faa3092e5595        4 minutes ago       675MB
+node                          12-alpine           08a8c8089ab1        2 weeks ago         673MB
+```
+
+# Step 4 - Run a Container
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:209]: This needs to be a H2 heading

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+
+# Step 1 - Creating a Node.js Application
+
+First, we will start by creating a directory for our project then install some dependencies for our simple Hello World Website.
+```
+ mkdir node-web-app
+ cd node-web-app
+```
+
+### Install NPM and Express Framework
```
[the remark this answers is not in the document]
Comment (louisefindlay23) [reply articles/building-a-nodejs-application-using-docker/index.md]: This issue needs to be fixed, @damikanye. I've marked it with new comments.

**Comment by damikanye** [comment]
@louisefindlay23 have made the changes

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Thanks for making the changes, @damikanye. Just a few minor things from your revisions. Correct them and I'll do another full review.

**Comment by damikanye** [comment]
Made some minor changes @louisefindlay23

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 20]
```diff
@@ -0,0 +1,244 @@
+# Building A Node.js Application Using Docker
+
+[Docker](https://www.docker.com/) allows us to run our applications as *containers*. A container is a standalone executable package which is lightweight and has everything needed to run an application be it libraries, tools, runtime, setting or code.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:3]: setting**s**

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+
+
+## Dockerize the Node.js Application
+
+To start this tutorial, we need to make sure Docker and [Node.js](https://nodejs.org/en/download/package-manager/) is installed on our system. If Docker is not, you can refer to the links below to download:
+- [Ubuntu](https://docs.docker.com/engine/installation/linux/ubuntu/)
+- [Windows](https://docs.docker.com/docker-for-windows/install/)
+- [MacOS](https://docs.docker.com/docker-for-mac/install/)
+
+In order to dockerize a Node.js Application we need to follow these steps:
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:15]: Lowercase application and add a comma afterwards.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+
+- Creating A Node.js Application
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:17]: lowercase application

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+- Create A DockerFile
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:18]: Lowercase f in Dockerfile

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+- Building your Docker Image
+- Expose
+
+## Step 1 - Creating a Node.js Application
+
+First, we will start by creating a directory for our project then install some dependencies for our simple Hello World Website.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:24]: add and before then or a full stop. Lowercase website

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+```bash
+ mkdir node-web-app
+ cd node-web-app
+```
+
+## Install npm And Express Framework
+
+Install npm and express framework, which is a Node.js framework. Then, initialize npm in our directory.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:32]: Please revise framework, which is a Node.js framework to flow better.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+
+# Copying rest of the application to app directory
+COPY . /app
+ 
+# Expose the port and start the application
+Expose 8080
+
+CMD ["npm","start"]
+```
+*NB: Above you will notice we used two distinct COPY command to reduce the application rebuild time*
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:134]: command**s**

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+```
+This prevents the local module and debug logs from being copied onto your Docker image.
+
+## Step 3 - Building your Docker Image
+
+Building your Docker image is quite easy and can be done using a single command.
+```
+docker build -t <docker-image-name> <filepath>
+```
+*The `-t` flag lets you tag your image so its easier to find later*
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:151]: it**’s**

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+
+For Example:
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:153]: lowercase example

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+Step 7/7 : CMD ["npm","start"]
+ ---> Running in 50a5824edf5e
+Removing intermediate container 50a5824edf5e
+ ---> faa3092e5595
+Successfully built faa3092e5595
+Successfully tagged cisca:latest
+```
+You should get something similar to the output above after executing the command. This means that the docker image was created successfully and the app is working fine.
+
+Now that the build is complete you can check your image:
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:198]: Add comma after complete.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 30]
```diff
@@ -0,0 +1,244 @@
+# Building A Node.js Application Using Docker
+
+[Docker](https://www.docker.com/) allows us to run our applications as *containers*. A container is a standalone executable package which is lightweight and has everything needed to run an application be it libraries, tools, runtime, setting or code.
+
+In this tutorial, we will build a Node.js application, create an image and also build a container using the image. Enjoy!
+
+
+## Dockerize the Node.js Application
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:8]: Still need to fix the MD headings as per review.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Thanks, @damikanye for making the changes. Please fix the Markdown headings though.

**Comment by louisefindlay23** [comment]
> Plagiarism check done

Thanks, @hectorkambow. Could you add some front-matter please?

**Review by louisefindlay23 — APPROVED** [review]
Thanks for making the suggested changes, @damikanye. Your article has passed preliminary revision and on to the tech review.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+
+The app is ready to launch:
+```
+$ node app.js
+```
+Go to `http://localhost:8080/` in your browser to view it.
+
+## Step 2 - Create A DockerFile
+
+Create a file in the directory called `Dockerfile`.
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:86]: I would change this to root directory so readers don't think Dockerfile should be a directory.

File: articles/building-a-nodejs-application-using-docker/index.md [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+
+Now that the build is complete, you can check your image:
+```
+docker image ls
+```
+Result:
+```
+REPOSITORY                    TAG                 IMAGE ID            CREATED             SIZE
+cisca                         latest              faa3092e5595        4 minutes ago       675MB
+node                          12-alpine           08a8c8089ab1        2 weeks ago         673MB
```
Comment (louisefindlay23) [inline articles/building-a-nodejs-application-using-docker/index.md:206]: My tag for the docker container was 9 but I assume that's variances in installation.

**Review by louisefindlay23 — APPROVED** [review]
Great work, @damikanye. Just one minor suggestion and your article is ready for publication.

File: articles/building-a-nodejs-application-using-docker/index.md (hunk 32 above)
Comment (damikanye) [reply articles/building-a-nodejs-application-using-docker/index.md:206]: Yes

**Comment by damikanye** [comment]
> Great work, @damikanye. Just one minor suggestion and your article is ready for publication.

Thanks, @louisefindlay23. Have made the minor changes.

**Comment by louisefindlay23** [comment]
Thanks, @damikanye. A great article that our readers are sure to love and find useful. I certainly found it informative since I've used Node.js and Docker before, but never in combination.

@hectorkambow, this article is ready for publication once front-matter has been added.

**Pull request merged**